### PR TITLE
periodic reauth

### LINKED ISSUE #2831: 802.1x reauth interval
We would like to be able to re-auth a port periodically, even if the port doesn't change op status.

### PR DESCRIPTION
**Pull request by Bairdo**
closes #2831 

Builds on top of PR #2856 

Requires https://github.com/faucetsdn/chewie/pull/117

### CODE REVIEW

File: tests/integration/mininet_tests.py [hunk 1]
```diff
@@ -251,16 +253,35 @@ def setUp(self):
         super(Faucet8021XSuccessTest, self).setUp()
         self.host_drop_all_ips(self.nfv_host)
         self.radius_log_path = self.start_freeradius()
+        self.eapol1_host.cmd('tcpdump -w %s/%s-start.pcap ether proto 0x888e &' %
```
Comment (Bairdo) [inline tests/integration/mininet_tests.py]: tcpdumps are only used for debugging.

**Comment by Bairdo** [comment]
haven't tested on hardware yet (running now)

**Comment by Bairdo** [comment]
x510 passes.
Cisco fails (Faucet9021XIdentityOnPortUpTest, Faucet8021XPeriodicReauthTest, Faucet8021XPortFlapTest, Faucet8021XSuccessTest)

**Comment by Bairdo** [comment]
Just reran on the cisco (without any changes) a few more times and different tests fail each time.

Run 1: failed - Faucet8021XIdentityOnPortUpTest and Faucet8021XPeriodicReauthTest

Run 2: failed - Faucet8021XIdentityOnPortUpTest, Faucet8021XPeriodicReauthTest, Faucet8021XSuccessTest

Run 3: failed - Faucet8021XIdentityOnPortUpTest, Faucet8021XPeriodicReauthTest, Faucet8021XPortStatusTest, Faucet8021XSuccessTest

Run 4 (increased the port up timer to 40sec instead of 20): failed -  Faucet8021XFailureTest, Faucet8021XIdentityOnPortUpTest, Faucet8021XPeriodicReauthTest, Faucet8021XPortFlapTest

**Comment by Bairdo** [comment]
Note to self:
run 3 /var/tmp/faucet-tests-w2qu9zxh
```
======================================================================
FAIL: test_untagged (mininet_tests.Faucet8021XIdentityOnPortUpTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/faucet-src/tests/integration/mininet_tests.py", line 672, in test_untagged
    self.assertIn("len 5, Request (1)", tcpdump_txt)
AssertionError: 'len 5, Request (1)' not found in '0 packets captured0 packets received by filter0 packets dropped by kernel'

======================================================================
FAIL: test_untagged (mininet_tests.Faucet8021XPeriodicReauthTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/faucet-src/tests/integration/mininet_tests.py", line 712, in test_untagged
    self.scrape_prometheus_var('port_dot1x_success_total', labels=port_labels1, default=0))
AssertionError: 4 != 1

======================================================================
FAIL: test_untagged (mininet_tests.Faucet8021XPortStatusTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/faucet-src/tests/integration/mininet_tests.py", line 601, in test_untagged
    'port_dot1x_success_total', labels=self.port_labels(port_no1), default=0))
AssertionError: 1 != 2

======================================================================
FAIL: test_untagged (mininet_tests.Faucet8021XSuccessTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/faucet-src/tests/integration/mininet_tests.py", line 333, in test_untagged
    terminate_wpasupplicant=True)
  File "/faucet-src/tests/integration/mininet_tests.py", line 285, in try_8021x
    timeout=15, vflags='-vvv', packets=10)
  File "/faucet-src/clib/mininet_test_base.py", line 663, in tcpdump_helper
    return TcpdumpHelper(*args, **kwargs).execute()
  File "/faucet-src/clib/tcpdump_helper.py", line 80, in execute
    line = self.next_line()
  File "/faucet-src/clib/tcpdump_helper.py", line 144, in next_line
    func()
  File "/faucet-src/tests/integration/mininet_tests.py", line 284, in <lambda>
    terminate_wpasupplicant=terminate_wpasupplicant)],
  File "/faucet-src/tests/integration/mininet_tests.py", line 394, in wpa_supplicant_callback
    self.wait_for_eap_success(host, wpa_ctrl_path)
  File "/faucet-src/tests/integration/mininet_tests.py", line 429, in wait_for_eap_success
    self.fail('did not get EAP success: %s' % eap_state)
AssertionError: did not get EAP success: IDLE

----------------------------------------------------------------------
```

**Comment by Bairdo** [comment]
ofchannel.txt, shows the identity request being packet in-ed. But it should hit a redirect ACL.

Printing the flows from the switch console. shows that the ACL is not there when the packet is packet in-ed.
But why's it not there?
- it's deleted on port down. 
- should be re added (we see the flow add in ofchannel.txt before the packet is sent) on port up.

t=00:24:56 - port down
t=00:25:01 - port up
t=00:25:01 - add flows
t=00:25:41 - packet in

```
Mar 21 00:25:41 /var/tmp/faucet-tests-gynvauwn/Faucet8021XIdentityOnPortUpTest-test_untagged/ofchannel.txt DEBUG    1/1 DPID 46391479553 (0xacd25cd01) version=0x4,msg_type=0xa,msg_len=0x6a,xid=0x0,OFPPacketIn(buffer_id=258,cookie=12398534311923819730,data=b'\x01\x80\xc2\x00\x00\x03\x00\x00\x00\x00\x00\x05\x81\x00\x00d\x88\x8e\x01\x00\x00\x05\x01\x08\x00\x05\x01\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',match=OFPMatch(oxm_fields={'in_port': 17}),reason=1,table_id=2,total_len=64)
Mar 21 00:25:41 /var/tmp/faucet-tests-gynvauwn/Faucet8021XIdentityOnPortUpTest-test_untagged/ofchannel.txt DEBUG    1/2 DPID 46391479553 (0xacd25cd01) version=None,msg_type=None,msg_len=None,xid=None,OFPFlowMod(buffer_id=4294967295,command=0,cookie=12398534311923819730,cookie_mask=0,flags=0,hard_timeout=0,idle_timeout=429,instructions=[OFPInstructionActions(actions=[OFPActionPopVlan(len=8,type=18), OFPActionOutput(len=16,max_len=0,port=17,type=0)],type=4)],match=OFPMatch(oxm_fields={'vlan_vid': 4196, 'eth_dst': '00:00:00:00:00:05'}),out_group=0,out_port=0,priority=8192,table_id=3)
Mar 21 00:25:41 /var/tmp/faucet-tests-gynvauwn/Faucet8021XIdentityOnPortUpTest-test_untagged/ofchannel.txt DEBUG    2/2 DPID 46391479553 (0xacd25cd01) version=None,msg_type=None,msg_len=None,xid=None,OFPFlowMod(buffer_id=4294967295,command=0,cookie=12398534311923819730,cookie_mask=0,flags=0,hard_timeout=279,idle_timeout=0,instructions=[OFPInstructionGotoTable(len=8,table_id=3,type=1)],match=OFPMatch(oxm_fields={'eth_src': '00:00:00:00:00:05', 'vlan_vid': 4196, 'in_port': 17}),out_group=0,out_port=0,priority=8191,table_id=2)
```

**Comment by Bairdo** [comment]
Problem was with two flows on the nfv port having the same priority value, but different match conditions (undefined behaviour). A little odd that it would work when the supplicant initiated. But that's my fault.

OVS, CIsco pass, (running x510 now)

**Comment by Bairdo** [comment]
So this PR was branched off of PR #2856 , but contains a bug or two which have been fixed in this PR. So either they can both be merged in or #2856 can be deleted as it is contained in this one, without the errors.

**Comment by Bairdo** [comment]
um 42 files changed?

**Comment by Bairdo** [comment]
I'm not sure whats going on with the merge. I've tried variations (on other tmp branches) of merging/rebasing with master, and going through to create a new PR, but all are trying to apply other changes.

**Comment by Bairdo** [comment]
Ok fixed this now.

**Comment by gizmoguy** [comment]
Tests are failing due because this PR needs https://github.com/faucetsdn/chewie/pull/117 merged and a new chewie release minted.

File: tests/unit/faucet/test_valve.py [hunk 2]
```diff
@@ -1518,12 +1518,13 @@ def test_get_mac_str(self):
...
         port_no = 1
+        vlan_name = 'student'
+        filter_id = 'block_http'
         for handler in (
-                self.dot1x.auth_handler,
                 self.dot1x.logoff_handler,
                 self.dot1x.failure_handler):
-            handler(
-                '0e:00:00:00:00:ff', faucet_dot1x.get_mac_str(valve_index, port_no))
+        self.dot1x.auth_handler(
```
Comment (gizmoguy) [inline tests/unit/faucet/test_valve.py]: Bad indentation here.

`33% 4:8=8s /home/travis/build/faucetsdn/faucet/tests/unit/faucet/test_valve.py  Couldn't run '/home/travis/build/faucetsdn/faucet/tests/unit/faucet/test_valve.py' as Python code: IndentationError: expected an indented block (test_valve.py, line 1526)`

**Pull request merged**